### PR TITLE
SONAR-7260 Fail fast if version of database is not supported

### PR DESCRIPTION
**Pull request by simonbrandhof**
See https://jira.sonarsource.com/browse/SONAR-7260

### CODE REVIEW

File: sonar-db/src/main/java/org/sonar/db/DatabaseChecker.java [hunk 1]
```diff
@@ -43,12 +44,29 @@ public DatabaseChecker(Database db) {
   @Override
   public void start() {
     try {
-      if (H2.ID.equals(db.getDialect().getId())) {
-        Loggers.get(DatabaseChecker.class).warn("H2 database should be used for evaluation purpose only");
-      } else if (Oracle.ID.equals(db.getDialect().getId())) {
-        checkOracleVersion();
+      switch (db.getDialect().getId()) {
```
Comment (SonarTech) [inline sonar-db/src/main/java/org/sonar/db/DatabaseChecker.java]: MAJOR Add a default case to this switch. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3ASwitchLastCaseIsDefaultCheck)

File: sonar-db/src/main/java/org/sonar/db/DatabaseChecker.java [hunk 2]
```diff
@@ -43,12 +44,29 @@ public DatabaseChecker(Database db) {
...
+          checkOracleDriverVersion();
+          checkMinDatabaseVersion(db.getDialect(), 11, 0);
+          break;
+        case MySql.ID:
+          checkMinDatabaseVersion(db.getDialect(), 5, 6);
+          break;
+        case PostgreSql.ID:
+          checkMinDatabaseVersion(db.getDialect(), 8, 0);
+          break;
+        case MsSql.ID:
```
Comment (SonarTech) [inline sonar-db/src/main/java/org/sonar/db/DatabaseChecker.java]: MAJOR Reduce this switch case number of lines from 6 to at most 5, for example by extracting code into methods. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1151)

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 issues:
* MAJOR 2 major

Watch the comments in this conversation to review them.

File: sonar-db/src/main/java/org/sonar/db/DatabaseChecker.java [hunk 3]
```diff
@@ -58,29 +77,49 @@ public void stop() {
...
+      this.minor = minor;
+    }
+
+    public boolean isGreaterThanOrEqual(Version other) {
+      return major >= other.major && (major != other.major || minor >= other.minor);
+    }
+
+    @Override
+    public String toString() {
+      return String.format("%d.%d", major, minor);
```
Comment (sns-seb) [inline sonar-db/src/main/java/org/sonar/db/DatabaseChecker.java]: as a general practice, `String.format` should be avoided in `toString()` methods, this method can be intensively called

It's not really the case, but here `major + '.' + minor` wouldn't be much longer to write nor harder to read

or did I miss a benefit of `String.format` in this case?

**Comment by sns-seb** [comment]
some picky comments, otherwise LGTM

File: sonar-db/src/main/java/org/sonar/db/DatabaseChecker.java [hunk 4]
```diff
@@ -20,19 +20,35 @@
...
 public class DatabaseChecker implements Startable {
 
-  public static final int ORACLE_MIN_MAJOR_VERSION = 11;
+  private static final Map<String, Version> MINIMAL_SUPPORTED_DB_VERSIONS = ImmutableMap.of(
+    // MsSQL 2008 is 10.x
+    // MsSQL 2012 is 11.x
+    // MsSQL 2014 is 12.x
+    // https://support.microsoft.com/en-us/kb/321185
+    MsSql.ID, new Version(10, 0),
+
```
Comment (sns-seb) [inline sonar-db/src/main/java/org/sonar/db/DatabaseChecker.java]: why a line return here?

**Pull request closed without merging**